### ISSUE TITLE
0 (zero) in apply coupon returns 'empty coupon code'

### ISSUE THREAD

**Issue by Arifursdev** [woothemes/woocommerce#35210]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

It allows to create discount code: 0
but when trying to apply '0' coupon in cart, it returns 

'Coupon "0" does not exist!'

possibly because for validating you have used PHP 'empty' function which returns true for 0
https://www.php.net/manual/en/function.empty.php#usernotes

### Expected behavior

It should have applied the coupon in the cart.

### Actual behavior

it returns 'Coupon "0" does not exist!' when trying to apply coupon code: 0 (zero)

### Steps to reproduce

1. Create coupon code: 0
2. and try to apply it on cart

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://localhost/wc_cart
Site address (URL): http://localhost/wc_cart
WC Version: 6.6.1
REST API Version: ✔ 6.6.1
WC Blocks Version: ✔ 7.6.2
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.0.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.48 (Win64) OpenSSL/1.1.1k PHP/8.0.8
PHP Version: 8.0.8
PHP Post Max Size: 40 MB
PHP Time Limit: 120
PHP Max Input Vars: 1000
cURL Version: 7.76.1
OpenSSL/1.1.1k

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.4.20-MariaDB
Max Upload Size: 40 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.6.1
WC Database Prefix: wp_
Total Database Size: 4.61MB
Database Data Size: 3.14MB
Database Index Size: 1.47MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.06MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.13MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 23
page: 7
post: 2
product: 18
product_variation: 7
shop_coupon: 3

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (3) ###

Loco Translate: by Tim Whitlock – 2.6.2
Cart Drawer: by dev– 2.0
WooCommerce: by Automattic – 6.6.1 (update to version 7.0.0 is available)

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: GeneratePress
Version: 3.1.0 (update to version 3.1.3 is available)
Author URL: https://tomusborne.com
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
wc-pay-promotion
wc-pay-welcome-page
wc-pay-subscriptions-page

Disabled Features: minified-js
settings

Daily Cron: ✔ Next scheduled: 2022-10-21 09:24:32 +00:00
Options: ✔
Notes: 54
Onboarding: skipped

### Action Scheduler ###

Complete: 4
Oldest: 2022-10-18 20:15:13 +0000
Newest: 2022-10-20 13:19:58 +0000

Pending: 1
Oldest: 2022-10-21 13:18:50 +0000
Newest: 2022-10-21 13:18:50 +0000

### Status report information ###

Generated at: 2022-10-20 13:55:38 +00:00
`

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by swatipawarGS** [comment]
Hi @Arifursdev,

Thank you for taking the time to report this bug, we really appreciate your help. We are able to reproduce the issue on our end using the  WooCommerce 7.0.0 version, PHP 8.0.24, and WordPress 6.0.3 version. But we get the 'Please enter a coupon code.' error instead of 'Coupon "0" does not exist!' error.

Please find below the screencast for reference :

https://user-images.githubusercontent.com/94531721/197128555-1f0a888b-9bcf-45b7-86ca-90d63b93be17.mp4


Priority of this issue will be assigned shortly after we’d go through a round of prioritization. Please note it may take a couple of days.

**Comment by Arifursdev** [comment]
Hi @swatipawarGS,

thanks for the screencast. that's correct, error I have submitted is coming from a custom plugin we have built for cart drawer. which picks up that error `Coupon $code does not exist!` from `wc_get_notices` function after trying to apply discount with `add_discount` function from `WC()->cart` .

its not a custom message, but i think its coming from WooCommerce and translation can be seen on `plugins\woocommerce\i18n\languages\woocommerce.pot`

// line starting from 19660
`#. translators: %s: coupon code`
`#: includes/class-wc-coupon.php:987`
`#: includes/class-wc-discounts.php:590`
`msgid "Coupon \"%s\" does not exist!"`
`msgstr ""`

the error you're seeing from cart page on any default theme is possibly from:
`class-wc-ajax.php > static function apply_coupon` 

please find below screenshot for reference
[https://i.ibb.co/LRBSf7m/Screenshot-16.png](url)
we're not using default ajax hook from wc for applying coupon

hope that makes sense.

**Comment by Konamiman** [comment]
Fix: https://github.com/woocommerce/woocommerce/pull/36924
